### DISCUSSION TITLE
How to create new module

### DISCUSSION THREAD

**Discussion by vicky-gonsalves** [Prolifode/deno_rest#61] [Q&A]
https://github.com/Prolifode/deno_rest/issues/60#issue-1989269750

**Comment by vicky-gonsalves** [comment]
In the project structure, there are 5 basic directories,
1. controllers
2. routers
3. services
4. models
5. validations

--------------------------------

1. `controllers` dir holds the controller of the module. For an example, to create new module named `article`, in `controllers` dir create new controller file named `article.controller.ts` and create new class named `ArticleController`

```
class ArticleController {
  public static async create(
    { request, response }: RouterContext<string>,
  ): Promise<void> {
    const body = request.body();
    ...
    ...
    response.status = Status.Created;
  }

}

```


2. `routers` dir holds the routes of the module. To create new route for POST article, create new file under `routers` dir named `article.router.ts` and add routes as per requirement. Check `user.router.ts` for reference. 
eg. To create new Article we would add new route :
```
router.post(
  '/api/articles',
  ArticleController.create,
);
```
if you want to protect the above route using auth, validation and permissions,
add the middleware :
```
router.post(
  '/api/articles',
  auth([PermissionList.MANAGE_ARTICLES]),   ---------> add this new permission in `config/roles.ts`
  validate(createUserValidation),                      ---------> create new validation under `validations` dir
  ArticleController.create,
);
```
Once done initialize this new router in `router/index.ts` file:
```

const init = (app: Application) => {
  app.use(authRouter.routes());
  app.use(userRouter.routes());
  app.use(defaultRouter.routes());  
  app.use(ArticleRouter.routes());   ---------------------> new router  
  ...
  app.use(ArticleRouter.allowedMethods());    ---------------------> new router  

```


3. services is the main heart of the module where all processing happens. We use Thin controller and Fat Services approach. To create a new service, add new service file `article.service.ts` under `services` dir.
```
class ArticleService {  
  public static async createArticle(
    options: CreateArticleStructure,
  ): Promise<string | Bson.ObjectId | Error> {
    // add your service logic
    }
    
```


4. models contains the database schema. So to create new schema for article, add new file `article.model.ts` under `models` dir.
 
5. Similarly you can add route specific validations in validations files.

**Comment by I-did-this** [comment]
Hi @vicky-gonsalves,

I successfully deployed a product based on the deno_rest project. However, I am currently stuck with the Oak library and need to upgrade it. I could use your help with this.

I have also emailed you regarding the same on this email [vicky.gonsalves@outlook.com](mailto:vicky.gonsalves@outlook.com). Let me know when you have time to connect.

Thanks!

**Reply by vicky-gonsalves** [reply]
@nehaljadav Could you describe the issue in more detail? Sharing the exact error messages would help me understand the problem better. If possible, please attach screenshots or the full error stack for better clarity.

**Reply by vicky-gonsalves** [reply]
@nehaljadav I have added support for deno 2.x and  There are many breaking changes. Please go through this PR and update your project https://github.com/Prolifode/deno_rest/pull/66

**Reply by I-did-this** [reply]
Hi @vicky-gonsalves, thanks for your kind response. I see there are significant changes. Let me update my project accordingly. I will get back to you soon with my next question.
